src/gui/TrackContentObjectView.cpp

```diff
@@ -1106,6 +1127,100 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
+{
+	// Can't merge a single TCO
+	if (tcovs.size() < 2) { return false; }
+
+	// We check if the owner of the first TCO is an Instrument Track
+	bool isInstrumentTrack = dynamic_cast<InstrumentTrackView*>(tcovs.at(0)->getTrackView());
+
+	// Then we create a set with all the TCOs owners
+	std::set<TrackView*> ownerTracks;
+	for (auto tcov: tcovs) { ownerTracks.insert(tcov->getTrackView()); }
+
+	// Can merge if there's only one owner track and it's an Instrument Track
+	return isInstrumentTrack && ownerTracks.size() == 1;
+}
+
+void TrackContentObjectView::mergeTCOs(QVector<TrackContentObjectView*> tcovs)
+{
+	// Get the track that we are merging TCOs in
+	InstrumentTrack* track =
+		dynamic_cast<InstrumentTrack*>(tcovs.at(0)->getTrackView()->getTrack());
+
+	if (!track)
+	{
+		qWarning("Warning: Couldn't retrieve InstrumentTrack in mergeTCOs()");
+		return;
+	}
+
+	// For Undo/Redo
+	track->addJournalCheckPoint();
+	track->saveJournallingState(false);
+
+	// Find the earliest position of all the selected TCOVs
+	const auto earliestTCOV = std::min_element(tcovs.constBegin(), tcovs.constEnd(),
+		[](TrackContentObjectView* a, TrackContentObjectView* b)
+		{
+			return a->getTrackContentObject()->startPosition() <
+				b->getTrackContentObject()->startPosition();
+		}
+	);
+
+	const TimePos earliestPos = (*earliestTCOV)->getTrackContentObject()->startPosition();
+
+	// Create a pattern where all notes will be added
+	Pattern* newPattern = dynamic_cast<Pattern*>(track->createTCO(earliestPos));
+	if (!newPattern)
+	{
+		qWarning("Warning: Failed to convert TCO to Pattern on mergeTCOs");
+		return;
+	}
+
+	newPattern->saveJournallingState(false);
+
+	// Add the notes and remove the TCOs that are being merged
+	for (auto tcov: tcovs)
+	{
+		// Convert TCOV to PatternView
+		PatternView* pView = dynamic_cast<PatternView*>(tcov);
+
+		if (!pView)
+		{
+			qWarning("Warning: Non-pattern TCO on InstrumentTrack");
+			continue;
+		}
+
+		NoteVector currentTCONotes = pView->getPattern()->notes();
+		TimePos pViewPos = pView->getPattern()->startPosition();
+
+		for (Note* note: currentTCONotes)
+		{
+			Note* newNote = newPattern->addNote(*note, false);
+			TimePos originalNotePos = newNote->pos();
+			newNote->setPos(originalNotePos + (pViewPos - earliestPos));
+		}
+
+		// We disable the journalling system before removing, so the
+		// removal doesn't get added to the undo/redo history
+		tcov->getTrackContentObject()->saveJournallingState(false);
+		// No need to check for nullptr because we check while building the tcovs QVector
+		tcov->remove();
+	}
+
+	// Update length since we might have moved notes beyond the end of the pattern length
+	newPattern->updateLength();
+	// Rearrange notes because we might have moved them
+	newPattern->rearrangeAllNotes();
+	// Restore journalling states now that the operation is finished
+	track->restoreJournallingState();
+	newPattern->restoreJournallingState();
```

Comment:
Not important, but I want to change the order of two calls because `saveJournallingState`/`restoreJournallingState` works like push/pop.
```suggestion
	newPattern->restoreJournallingState();
	track->restoreJournallingState();
```

Reply:
Done! Does that mean that when I called `track->restoreJournallingState();` I was actually restoring `newPattern`s journalling state?

Reply:
It's rather a stylistic/semantic point, nothing was working wrong.

Reply:
Oh, got it now! Thanks!